library/src-bootstrapped/scala/runtime/Tuple.scala

```diff
@@ -463,4 +463,38 @@ object Tuple {
   def concatIterator(tup1: Tuple, tup2: Tuple): Iterator[Any] =
     tup1.asInstanceOf[Product].productIterator ++ tup2.asInstanceOf[Product].productIterator
 
+  def isInstanceOfTuple(x: Any): Boolean =
+    x.isInstanceOf[Product] && isInstanceOfEmptyTuple(x) || isInstanceOfNonEmptyTuple(x)
+
+  def isInstanceOfEmptyTuple(x: Any): Boolean = x == Tuple0
+
+  def isInstanceOfNonEmptyTuple(x: Any): Boolean = isInstanceOfPair(x)
+
+  def isInstanceOfPair(x: Any): Boolean =
+    x.isInstanceOf[Product] && (
+         x.isInstanceOf[Tuple1[_]]
```

Comment:
Can we make the test faster? Maybe cache all Tuple classes in a lazy val of type Set and then do a test with getClass? The initial `x.isInstanceOf[Product] ` is good, but even then I fear we get into the disjunction a lot. So whenever we have a Product class that is not a Tuple this test is going through all alternatives.

Reply:
An even faster alternative would be a switch on the product size